// @botonic/react/lib/webchat/webchatDev.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebchatDev = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _hooks = require("./hooks");

var _webchat = require("./webchat");

var _sessionView = require("./sessionView");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var WebchatDev = (0, _react.forwardRef)(function (props, ref) {
  var webchatHooks = (0, _hooks.useWebchat)();
  var webchatState = webchatHooks.webchatState,
      updateTheme = webchatHooks.updateTheme;
  /* TODO: webchatState.theme shoud be included in the dependencies array
  together with props.theme. The problem is that this effect modifies webchatState
  so we enter an infinite rerender loop. */

  (0, _react.useEffect)(function () {
    return updateTheme(_objectSpread({}, webchatState.theme, {}, props.theme));
  }, [props.theme]);
  return _react["default"].createElement("div", null, _react["default"].createElement(_webchat.Webchat, (0, _extends2["default"])({
    style: {
      flex: 1,
      position: 'relative'
    }
  }, props, {
    ref: ref,
    webchatHooks: webchatHooks,
    initialSession: {
      is_first_interaction: true,
      last_session: {},
      user: {
        id: '000001',
        username: 'johndoe',
        name: 'John Doe',
        provider: 'dev',
        provider_id: '0000000',
        extra_data: {}
      },
      organization: '',
      bot: {
        id: '0000000',
        name: 'botName'
      }
    },
    initialDevSettings: {
      keepSessionOnReload: webchatState.devSettings.keepSessionOnReload,
      showSessionView: webchatState.devSettings.showSessionView
    }
  })), _react["default"].createElement("div", {
    style: {
      position: 'fixed',
      left: 0,
      top: 0,
      width: 350,
      height: '100%'
    }
  }, _react["default"].createElement(_sessionView.SessionView, {
    webchatHooks: webchatHooks
  })));
});
exports.WebchatDev = WebchatDev;
