// @botonic/react/lib/webchat/header.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebchatHeader = exports.DefaultHeader = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _contexts = require("../contexts");

var _utils = require("../utils");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _botonic_react_logo100x = _interopRequireDefault(require("./botonic_react_logo100x100.png"));

var _grid = require("@rebass/grid");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  flex-direction: column;\n  justify-content: center;\n  flex: 1 1 auto;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 10px;\n  align-items: center;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 0px 16px;\n  cursor: pointer;\n  color: white;\n  font-family: sans-serif;\n  font-size: 36px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  background: linear-gradient(90deg, #2e203b 0%, ", " 100%);\n  height: 55px;\n  border-radius: 6px 6px 0px 0px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  @import url('https://fonts.googleapis.com/css?family=Noto+Sans+JP');\n  font-family: Arial, Helvetica, sans-serif;\n  font-size: 11px;\n  color: #ffffff;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  @import url('https://fonts.googleapis.com/css?family=Noto+Sans+JP');\n  font-family: Arial, Helvetica, sans-serif;\n  font-size: 15px;\n  font-weight: bold;\n  color: #ffffff;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var HeaderTitle = (0, _styledComponents["default"])(_grid.Flex)(_templateObject());
var Subtitle = (0, _styledComponents["default"])(_grid.Flex)(_templateObject2());
var Diffuse = (0, _styledComponents["default"])(_grid.Flex)(_templateObject3(), function (props) {
  return props.color;
});

var CloseHeader = _styledComponents["default"].div(_templateObject4());

var StyledHeaderImage = (0, _styledComponents["default"])(_grid.Flex)(_templateObject5());
var StyledHeaderTitle = (0, _styledComponents["default"])(_grid.Flex)(_templateObject6());

var DefaultHeader = function DefaultHeader(props) {
  var HeaderImage = _botonic_react_logo100x["default"];
  if ('brandImage' in props.webchatState.theme) HeaderImage = props.webchatState.theme.brandImage;
  if ('headerImage' in props.webchatState.theme) HeaderImage = props.webchatState.theme.headerImage;
  var headerTitle = props.webchatState.theme.headerTitle;
  headerTitle = headerTitle || 'Botonic';
  var headerSubtitle = props.webchatState.theme.headerSubtitle;
  return _react["default"].createElement(Diffuse, {
    color: props.color
  }, HeaderImage && _react["default"].createElement(StyledHeaderImage, null, _react["default"].createElement("img", {
    style: {
      width: 32,
      borderRadius: '50%'
    },
    src: (0, _utils.staticAsset)(HeaderImage)
  })), _react["default"].createElement(StyledHeaderTitle, {
    ml: HeaderImage ? 0 : 16
  }, _react["default"].createElement(HeaderTitle, {
    mb: headerSubtitle ? '6px' : '0px'
  }, headerTitle), _react["default"].createElement(Subtitle, null, headerSubtitle)), _react["default"].createElement(CloseHeader, {
    onClick: props.onCloseClick
  }, "\u2A2F"));
};

exports.DefaultHeader = DefaultHeader;

var WebchatHeader = function WebchatHeader(props) {
  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      webchatState = _useContext.webchatState;

  var handleCloseWebchat = function handleCloseWebchat(event) {
    props.onCloseClick(event.target.value);
  };

  if (webchatState.theme.customHeader) {
    var CustomHeader = webchatState.theme.customHeader;
    return _react["default"].createElement(CustomHeader, null);
  }

  return _react["default"].createElement(DefaultHeader, {
    webchatState: webchatState,
    color: webchatState.theme.brandColor,
    onCloseClick: handleCloseWebchat
  });
};

exports.WebchatHeader = WebchatHeader;
