// @botonic/react/lib/webchat/webview.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebviewContainer = exports.WebviewHeader = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactFrameComponent = _interopRequireDefault(require("react-frame-component"));

var _contexts = require("../contexts");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var WebviewHeader = function WebviewHeader(props) {
  var _useContext = (0, _react.useContext)(_contexts.RequestContext),
      closeWebview = _useContext.closeWebview;

  return _react["default"].createElement("div", {
    style: {
      textAlign: 'right',
      backgroundColor: '#f4f4f4',
      borderTop: '1px solid rgba(0, 0, 0, 0.2)',
      borderBottom: '1px solid rgba(0, 0, 0, 0.2)'
    }
  }, _react["default"].createElement("div", {
    style: {
      display: 'inline-block',
      padding: '8px 12px',
      cursor: 'pointer'
    },
    onClick: closeWebview
  }, "\u2715"));
};

exports.WebviewHeader = WebviewHeader;

var WebviewContainer = function WebviewContainer(props) {
  var _useContext2 = (0, _react.useContext)(_contexts.WebchatContext),
      webchatState = _useContext2.webchatState;

  var _useContext3 = (0, _react.useContext)(_contexts.RequestContext),
      closeWebview = _useContext3.closeWebview;

  var Webview = webchatState.webview;

  var close = function close(e) {
    return e.data == 'botonicCloseWebview' && closeWebview();
  };

  (0, _react.useEffect)(function () {
    if (window.addEventListener) {
      window.addEventListener('message', close, false);
    } else if (window.attachEvent) {
      // ie8
      window.attachEvent('onmessage', close);
    }
  }, []);
  return _react["default"].createElement("div", {
    style: _objectSpread({
      position: 'absolute',
      display: 'flex',
      flexDirection: 'column',
      bottom: 0,
      width: '100%',
      height: '80%',
      backgroundColor: '#fff'
    }, props.style || {})
  }, _react["default"].createElement(WebviewHeader, {
    style: {
      flex: 'none'
    }
  }), _react["default"].createElement("div", {
    style: {
      flex: 1,
      overflow: 'auto'
    }
  }, typeof Webview === 'string' ? _react["default"].createElement("iframe", {
    src: Webview,
    style: {
      borderStyle: 'none',
      width: '100%',
      height: '100%'
    }
  }) : _react["default"].createElement(_reactFrameComponent["default"], {
    style: {
      borderStyle: 'none',
      width: '100%',
      height: '100%'
    }
  }, _react["default"].createElement(Webview, null))));
};

exports.WebviewContainer = WebviewContainer;
