// @botonic/react/lib/contexts.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebchatContext = exports.RequestContext = void 0;

var _react = _interopRequireDefault(require("react"));

var _hooks = require("./webchat/hooks");

var RequestContext = _react["default"].createContext({
  getString: function getString() {
    return '';
  },
  setLocale: function setLocale() {
    return '';
  },
  session: {},
  params: {},
  input: {},
  defaultDelay: 0,
  defaultTyping: 0
});

exports.RequestContext = RequestContext;

var WebchatContext = _react["default"].createContext({
  sendText: function sendText(text) {
    return '';
  },
  sendPayload: function sendPayload(payload) {
    return '';
  },
  setReplies: function setReplies(replies) {
    return '';
  },
  openWebview: function openWebview(webviewComponent) {
    return '';
  },
  addMessage: function addMessage(message) {
    return '';
  },
  updateReplies: function updateReplies(replies) {
    return '';
  },
  closeWebview: function closeWebview() {
    return '';
  },
  theme: {},
  webchatState: _hooks.webchatInitialState
});

exports.WebchatContext = WebchatContext;
