// @botonic/react/lib/webview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebviewApp = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _axios = _interopRequireDefault(require("axios"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = require("react-dom");

var _reactRouterDom = require("react-router-dom");

var _core = require("@botonic/core");

var _contexts = require("./contexts");

var App =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(App, _React$Component);

  function App(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, App);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(App).call(this, props));
    var url = new URL(window.location.href);
    var params = Array.from(url.searchParams.entries()).filter(function (_ref) {
      var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
          key = _ref2[0],
          value = _ref2[1];

      return key != 'context';
    }).reduce(function (o, _ref3) {
      var _ref4 = (0, _slicedToArray2["default"])(_ref3, 2),
          key = _ref4[0],
          value = _ref4[1];

      o[key] = value;
      return o;
    }, {});
    var session = JSON.parse(url.searchParams.get('context') || {});
    _this.state = {
      session: session,
      params: params
    };
    return _this;
  }

  (0, _createClass2["default"])(App, [{
    key: "close",
    value: function () {
      var _close = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee(options) {
        var payload, s, base_url, resp;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                payload = options ? options.payload : null;
                if (options.path) payload = "__PATH_PAYLOAD__".concat(options.path);

                if (!payload) {
                  _context.next = 15;
                  break;
                }

                if (options.params) {
                  payload = "".concat(payload, "?").concat((0, _core.params2queryString)(options.params));
                }

                s = this.state.session;
                _context.prev = 5;
                base_url = s._hubtype_api || 'https://api.hubtype.com';
                _context.next = 9;
                return (0, _axios["default"])({
                  method: 'post',
                  url: "".concat(base_url, "/v1/bots/").concat(s.bot.id, "/send_postback/"),
                  data: {
                    payload: payload,
                    chat_id: s.user.id
                  }
                });

              case 9:
                resp = _context.sent;
                _context.next = 15;
                break;

              case 12:
                _context.prev = 12;
                _context.t0 = _context["catch"](5);
                console.log(_context.t0);

              case 15:
                if (this.state.session.user.provider === 'whatsappnew') {
                  location.href = 'https://wa.me/' + this.state.session.user.imp_id;
                } else {
                  try {
                    window.WebviewSdk.close(function () {}, function (err) {
                      return console.log(err);
                    });
                  } catch (e) {}

                  try {
                    parent.postMessage('botonicCloseWebview', '*');
                  } catch (e) {}
                }

              case 16:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[5, 12]]);
      }));

      function close(_x) {
        return _close.apply(this, arguments);
      }

      return close;
    }()
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var requestContext = {
        getString: function getString(stringId) {
          return (0, _core.getString)(_this2.props.locales, _this2.state.session.__locale, stringId);
        },
        session: this.state.session || {},
        params: this.state.params || {},
        closeWebview: this.close.bind(this)
      };
      return _react["default"].createElement(_contexts.RequestContext.Provider, {
        value: requestContext
      }, this.props.webviews.map(function (Webview, i) {
        return _react["default"].createElement(_reactRouterDom.Route, {
          key: i,
          path: "/".concat(Webview.name),
          component: Webview
        });
      }));
    }
  }]);
  return App;
}(_react["default"].Component);

var WebviewApp =
/*#__PURE__*/
function () {
  function WebviewApp(_ref5) {
    var webviews = _ref5.webviews,
        locales = _ref5.locales;
    (0, _classCallCheck2["default"])(this, WebviewApp);
    this.webviews = webviews;
    this.locales = locales;
  }

  (0, _createClass2["default"])(WebviewApp, [{
    key: "render",
    value: function render(dest) {
      (0, _reactDom.render)(_react["default"].createElement(_reactRouterDom.BrowserRouter, null, _react["default"].createElement(App, {
        webviews: this.webviews,
        locales: this.locales
      })), dest);
    }
  }]);
  return WebviewApp;
}();

exports.WebviewApp = WebviewApp;
