// @botonic/react/lib/dev-app.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DevApp = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = require("react-dom");

var _reactBot = require("./react-bot");

var _webchatApp = require("./webchat-app");

var _webchat = require("./webchat");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var DevApp =
/*#__PURE__*/
function (_WebchatApp) {
  (0, _inherits2["default"])(DevApp, _WebchatApp);

  function DevApp(_ref) {
    var _this;

    var _ref$theme = _ref.theme,
        theme = _ref$theme === void 0 ? {} : _ref$theme,
        persistentMenu = _ref.persistentMenu,
        blockInputs = _ref.blockInputs,
        emojiPicker = _ref.emojiPicker,
        onInit = _ref.onInit,
        onOpen = _ref.onOpen,
        onClose = _ref.onClose,
        onMessage = _ref.onMessage,
        botOptions = (0, _objectWithoutProperties2["default"])(_ref, ["theme", "persistentMenu", "blockInputs", "emojiPicker", "onInit", "onOpen", "onClose", "onMessage"]);
    (0, _classCallCheck2["default"])(this, DevApp);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(DevApp).call(this, {
      theme: theme,
      persistentMenu: persistentMenu,
      blockInputs: blockInputs,
      emojiPicker: emojiPicker,
      onInit: onInit,
      onOpen: onOpen,
      onClose: onClose,
      onMessage: onMessage
    }));
    _this.bot = new _reactBot.ReactBot(_objectSpread({}, botOptions));
    return _this;
  }

  (0, _createClass2["default"])(DevApp, [{
    key: "render",
    value: function render(dest) {
      var _this2 = this;

      var optionsAtRuntime = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var _optionsAtRuntime$the = optionsAtRuntime.theme,
          theme = _optionsAtRuntime$the === void 0 ? {} : _optionsAtRuntime$the,
          persistentMenu = optionsAtRuntime.persistentMenu,
          blockInputs = optionsAtRuntime.blockInputs,
          emojiPicker = optionsAtRuntime.emojiPicker,
          onInit = optionsAtRuntime.onInit,
          onOpen = optionsAtRuntime.onOpen,
          onClose = optionsAtRuntime.onClose,
          onMessage = optionsAtRuntime.onMessage,
          webchatOptions = (0, _objectWithoutProperties2["default"])(optionsAtRuntime, ["theme", "persistentMenu", "blockInputs", "emojiPicker", "onInit", "onOpen", "onClose", "onMessage"]);
      theme = _objectSpread({}, this.theme, {}, theme);
      persistentMenu = persistentMenu || this.persistentMenu;
      blockInputs = blockInputs || this.blockInputs;
      emojiPicker = emojiPicker || this.emojiPicker;
      this.onInit = onInit || this.onInit;
      this.onOpen = onOpen || this.onOpen;
      this.onClose = onClose || this.onClose;
      this.onMessage = onMessage || this.onMessage;
      (0, _reactDom.render)(_react["default"].createElement(_webchat.WebchatDev, (0, _extends2["default"])({
        ref: this.webchatRef
      }, webchatOptions, {
        theme: theme,
        persistentMenu: persistentMenu,
        blockInputs: blockInputs,
        emojiPicker: emojiPicker,
        getString: function getString(stringId, session) {
          return _this2.bot.getString(stringId, session);
        },
        setLocale: function setLocale(locale, session) {
          return _this2.bot.setLocale(locale, session);
        },
        onInit: function onInit() {
          return _this2.onInitWebchat.apply(_this2, arguments);
        },
        onOpen: function onOpen() {
          return _this2.onOpenWebchat.apply(_this2, arguments);
        },
        onClose: function onClose() {
          return _this2.onCloseWebchat.apply(_this2, arguments);
        },
        onUserInput: function onUserInput() {
          return _this2.onUserInput.apply(_this2, arguments);
        }
      })), dest);
    }
  }, {
    key: "onUserInput",
    value: function () {
      var _onUserInput = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee(_ref2) {
        var _this3 = this;

        var input, session, lastRoutePath, resp;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                input = _ref2.input, session = _ref2.session, lastRoutePath = _ref2.lastRoutePath;
                this.onMessage && this.onMessage(this, {
                  from: 'user',
                  message: input
                });
                _context.next = 4;
                return this.bot.input({
                  input: input,
                  session: session,
                  lastRoutePath: lastRoutePath
                });

              case 4:
                resp = _context.sent;
                this.onMessage && resp.response.map(function (r) {
                  return _this3.onMessage(_this3, {
                    from: 'bot',
                    message: r
                  });
                });
                this.webchatRef.current.addBotResponse(resp);

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function onUserInput(_x) {
        return _onUserInput.apply(this, arguments);
      }

      return onUserInput;
    }()
  }]);
  return DevApp;
}(_webchatApp.WebchatApp);

exports.DevApp = DevApp;
