// @botonic/nlu/lib/constants.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.constants = mod.exports;
  }
})(this, function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.DEFAULT_HYPERPARAMETERS = _exports.DEFAULT_ENTITIES = _exports.GLOBAL_ENTITIES_REGEX = _exports.ENTITIES_REGEX = _exports.WORD_EMBEDDDINGS_ENDPOINT = _exports.DB = _exports.UNKNOWN_TOKEN = _exports.ASSETS_DIRNAME = _exports.UTTERANCES_EXTENSION = _exports.WORD_EMBEDDINGS_DIRNAME = _exports.GLOBAL_CONFIG_DIRNAME = _exports.NLU_CONFIG_FILENAME = _exports.NLU_DIRNAME = _exports.UTTERANCES_DIRNAME = _exports.MODELS_DIRNAME = _exports.MODEL_FILENAME = _exports.NLU_DATA_FILENAME = void 0;
  // Filenames
  var NLU_DATA_FILENAME = 'nlu-data.json';
  _exports.NLU_DATA_FILENAME = NLU_DATA_FILENAME;
  var MODEL_FILENAME = 'model.json'; // Dirnames

  _exports.MODEL_FILENAME = MODEL_FILENAME;
  var MODELS_DIRNAME = 'models';
  _exports.MODELS_DIRNAME = MODELS_DIRNAME;
  var UTTERANCES_DIRNAME = 'utterances'; // Subpaths

  _exports.UTTERANCES_DIRNAME = UTTERANCES_DIRNAME;
  var NLU_DIRNAME = 'nlu';
  _exports.NLU_DIRNAME = NLU_DIRNAME;
  var NLU_CONFIG_FILENAME = 'nlu.config.json';
  _exports.NLU_CONFIG_FILENAME = NLU_CONFIG_FILENAME;
  var GLOBAL_CONFIG_DIRNAME = '.botonic';
  _exports.GLOBAL_CONFIG_DIRNAME = GLOBAL_CONFIG_DIRNAME;
  var WORD_EMBEDDINGS_DIRNAME = 'word-embeddings'; // General Config

  _exports.WORD_EMBEDDINGS_DIRNAME = WORD_EMBEDDINGS_DIRNAME;
  var UTTERANCES_EXTENSION = '.txt';
  _exports.UTTERANCES_EXTENSION = UTTERANCES_EXTENSION;
  var ASSETS_DIRNAME = 'assets';
  _exports.ASSETS_DIRNAME = ASSETS_DIRNAME;
  var UNKNOWN_TOKEN = '<UNK>';
  _exports.UNKNOWN_TOKEN = UNKNOWN_TOKEN;
  var DB = {
    TABLE: 'embeddings',
    COLUMN: 'token'
  };
  _exports.DB = DB;
  var WORD_EMBEDDDINGS_ENDPOINT = 'https://s3-eu-west-1.amazonaws.com/word-embeddings.hubtype.com'; //Entities

  _exports.WORD_EMBEDDDINGS_ENDPOINT = WORD_EMBEDDDINGS_ENDPOINT;
  var ENTITIES_REGEX = /\[(.*?)\]\((.*?)\)/;
  _exports.ENTITIES_REGEX = ENTITIES_REGEX;
  var GLOBAL_ENTITIES_REGEX = /\[(.*?)\]\((.*?)\)/g;
  _exports.GLOBAL_ENTITIES_REGEX = GLOBAL_ENTITIES_REGEX;
  var DEFAULT_ENTITIES = [// Nouns
  'Organization', 'Currency', 'Unit', // Places
  'Country', 'Region', 'Place', 'City', // Dates
  'WeekDay', 'Date', 'Holiday', 'Month', 'Duration', 'Time', // People
  'FirstName', 'LastName', 'MaleName', 'FemaleName', 'Honorific', 'Person'];
  _exports.DEFAULT_ENTITIES = DEFAULT_ENTITIES;
  var DEFAULT_HYPERPARAMETERS = {
    EMBEDDING: '10k-fasttext',
    EMBEDDING_DIM: 300,
    TRAINABLE_EMBEDDINGS: true,
    LEARNING_RATE: 0.03,
    EPOCHS: 10,
    UNITS: 21,
    MAX_SEQ_LENGTH: 50,
    VALIDATION_SPLIT: 0.2,
    DROPOUT_REG: 0.2
  };
  _exports.DEFAULT_HYPERPARAMETERS = DEFAULT_HYPERPARAMETERS;
});
