// @botonic/react/lib/components/carousel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Carousel = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _message = require("./message");

var _core = require("@botonic/core");

var serialize = function serialize(carouselProps) {
  return {
    type: 'carousel',
    elements: carouselProps.children.map(function (e) {
      return e && e.type && e.type.serialize && e.type.serialize(e.props);
    })
  };
};

var Carousel = function Carousel(props) {
  var content = props.children;

  if ((0, _core.isBrowser)()) {
    content = _react["default"].createElement("div", {
      style: {
        paddingTop: 10,
        marginLeft: -13,
        display: 'flex',
        flexDirection: 'row',
        overflowX: 'auto',
        maxWidth: '100%'
      }
    }, _react["default"].createElement("div", {
      style: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'start'
      }
    }, props.children));
  }

  return _react["default"].createElement(_message.Message, (0, _extends2["default"])({
    style: {
      maxWidth: '85%',
      padding: 0,
      backgroundColor: 'transparent'
    },
    blob: false,
    json: serialize(props),
    type: "carousel"
  }, props), content);
};

exports.Carousel = Carousel;
Carousel.serialize = serialize;
