// @botonic/react/lib/components/pic.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pic = void 0;

var _react = _interopRequireDefault(require("react"));

var _core = require("@botonic/core");

var Pic = function Pic(props) {
  var renderBrowser = function renderBrowser() {
    return _react["default"].createElement("div", {
      style: {
        borderTopLeftRadius: 8,
        borderTopRightRadius: 8,
        width: 222,
        height: 140,
        background: "#fff url(".concat(props.src, ") no-repeat center/cover"),
        borderBottom: '1px solid #F1F0F0'
      }
    });
  };

  var renderNode = function renderNode() {
    return _react["default"].createElement("pic", null, props.src);
  };

  if ((0, _core.isBrowser)()) return renderBrowser();else if ((0, _core.isNode)()) return renderNode();
};

exports.Pic = Pic;

Pic.serialize = function (picProps) {
  return {
    pic: picProps.src
  };
};
