// @botonic/react/lib/components/document.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Document = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _message = require("./message");

var _core = require("@botonic/core");

var serialize = function serialize(documentProps) {
  return {
    document: documentProps.src
  };
};

var Document = function Document(props) {
  var content = props.children;
  if ((0, _core.isBrowser)()) content = _react["default"].createElement("embed", {
    style: {
      borderRadius: '8px',
      height: '300px',
      margin: '10px'
    },
    src: props.src
  });
  return _react["default"].createElement(_message.Message, (0, _extends2["default"])({
    json: serialize(props)
  }, props, {
    type: "document"
  }), content);
};

exports.Document = Document;
Document.serialize = serialize;
