// @botonic/react/lib/components/handoff.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Handoff = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _contexts = require("../contexts");

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  max-width: 60%;\n  padding: 12px 24px;\n  background-color: white;\n  border: none;\n  border-radius: 4px;\n  margin-top: 8px;\n  cursor: pointer;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  text-align: center;\n  white-space: normal;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  text-align: center;\n  white-space: normal;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  font-family: Arial, Helvetica, sans-serif;\n  flex-direction: column;\n  justify-content: center;\n  align-items: center;\n  padding: 24px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var Container = _styledComponents["default"].div(_templateObject());

var TransferredContainer = _styledComponents["default"].div(_templateObject2());

var EndedContainer = _styledComponents["default"].div(_templateObject3());

var StyledButton = _styledComponents["default"].button(_templateObject4());

var Handoff = function Handoff(props) {
  var _useContext = (0, _react.useContext)(_contexts.WebchatContext),
      resolveCase = _useContext.resolveCase;

  var _useState = (0, _react.useState)({
    showContinue: true
  }),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  var continueClick = function continueClick() {
    setState({
      showContinue: false
    });
    resolveCase();
  };

  var bgColor = state.showContinue ? '#c6e7c0' : '#d1d8cf';
  var fontColor = state.showContinue ? '#3a9c35' : '#5f735e';
  return _react["default"].createElement(Container, {
    style: {
      color: fontColor,
      backgroundColor: bgColor
    }
  }, state.showContinue ? _react["default"].createElement(TransferredContainer, null, "Conversation transferred to a human agent...") : _react["default"].createElement(EndedContainer, null, "Human handoff ended"), state.showContinue && _react["default"].createElement(StyledButton, {
    onClick: continueClick
  }, "Continue"));
};

exports.Handoff = Handoff;
