// @botonic/react/lib/components/image.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Image = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _message = require("./message");

var _core = require("@botonic/core");

var serialize = function serialize(imageProps) {
  return {
    image: imageProps.src
  };
};

var Image = function Image(props) {
  var content = props.children;
  if ((0, _core.isBrowser)()) content = _react["default"].createElement("img", {
    style: {
      borderRadius: '8px',
      maxWidth: '150px',
      maxHeight: '150px',
      margin: '10px'
    },
    src: props.src
  });
  return _react["default"].createElement(_message.Message, (0, _extends2["default"])({
    json: serialize(props)
  }, props, {
    type: "image"
  }), content);
};

exports.Image = Image;
Image.serialize = serialize;
