// @botonic/react/lib/components/location.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Location = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _core = require("@botonic/core");

var _message = require("./message");

var serialize = function serialize(locationProps) {
  return {
    location: {
      lat: locationProps.lat,
      "long": locationProps["long"]
    }
  };
};

var Location = function Location(props) {
  var renderBrowser = function renderBrowser() {
    var lat = parseFloat(props.lat);

    var _long = parseFloat(props["long"]);

    var location_url = "https://www.google.com/maps/@".concat(lat, ",").concat(_long);
    return _react["default"].createElement(_message.Message, (0, _extends2["default"])({
      json: serialize(props)
    }, props, {
      type: "location"
    }), _react["default"].createElement("a", {
      href: location_url,
      target: "_blank",
      rel: "noopener noreferrer"
    }, ' ', "See Location"));
  };

  var renderNode = function renderNode() {
    return _react["default"].createElement("message", {
      type: "location"
    }, _react["default"].createElement("lat", null, props.lat), _react["default"].createElement("long", null, props["long"]));
  };

  if ((0, _core.isBrowser)()) return renderBrowser();else if ((0, _core.isNode)()) return renderNode();
};

exports.Location = Location;
Location.serialize = serialize;
