// @botonic/react/lib/components/video.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Video = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _message = require("./message");

var _core = require("@botonic/core");

var serialize = function serialize(videoProps) {
  return {
    video: videoProps.src
  };
};

var Video = function Video(props) {
  var content = props.children;
  if ((0, _core.isBrowser)()) content = _react["default"].createElement("video", {
    style: {
      backgroundColor: 'rgba(0, 0, 0, 0.5)',
      borderRadius: '8px',
      maxHeight: '180px',
      maxWidth: '300px',
      margin: '10px'
    },
    controls: true
  }, _react["default"].createElement("source", {
    src: props.src
  }));
  return _react["default"].createElement(_message.Message, (0, _extends2["default"])({
    json: serialize(props)
  }, props, {
    type: "video"
  }), content);
};

exports.Video = Video;
Video.serialize = serialize;
