// @botonic/core/lib/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isFunction = isFunction;
exports.dirname = exports.join = exports.params2queryString = exports.isNode = exports.isBrowser = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _arguments = arguments;

var isBrowser = function isBrowser() {
  return typeof window !== 'undefined' && !window.process;
};

exports.isBrowser = isBrowser;

var isNode = function isNode() {
  return !isBrowser() && typeof process !== 'undefined';
};

exports.isNode = isNode;

function isFunction(o) {
  return typeof o === 'function';
}

var params2queryString = function params2queryString(params) {
  return Object.entries(params).map(function (_ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
        k = _ref2[0],
        v = _ref2[1];

    return "".concat(encodeURIComponent(k), "=").concat(encodeURIComponent(v));
  }).join('&');
}; // Joins path segments.  Preserves initial "/" and resolves ".." and "."
// Does not support using ".." to go above/outside the root.
// This means that join("foo", "../../bar") will not resolve to "../bar"


exports.params2queryString = params2queryString;

var join = function join()
/* path segments */
{
  // Split the inputs into a list of path commands.
  var parts = [];

  for (var i = 0, l = _arguments.length; i < l; i++) {
    parts = parts.concat(_arguments[i].split('/'));
  } // Interpret the path commands to get the new resolved path.


  var newParts = [];

  for (i = 0, l = parts.length; i < l; i++) {
    var part = parts[i]; // Remove leading and trailing slashes
    // Also remove "." segments

    if (!part || part === '.') continue; // Interpret ".." to pop the last segment

    if (part === '..') newParts.pop(); // Push new path segments.
    else newParts.push(part);
  } // Preserve the initial slash if there was one.


  if (parts[0] === '') newParts.unshift(''); // Turn back into a single string path.

  return newParts.join('/') || (newParts.length ? '/' : '.');
}; // A simple function to get the dirname of a path
// Trailing slashes are ignored. Leading slash is preserved.


exports.join = join;

var dirname = function dirname(path) {
  return join(path, '..');
};

exports.dirname = dirname;
